tomviz/EmdFormat.cxx

```diff
@@ -0,0 +1,461 @@
+/******************************************************************************
+
+  This source file is part of the tomviz project.
+
+  Copyright Kitware, Inc.
+
+  This source code is released under the New BSD License, (the "License").
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+******************************************************************************/
+#include "EmdFormat.h"
+
+#include "DataSource.h"
+
+#include <vtkDataArray.h>
+#include <vtkImageData.h>
+#include <vtkPointData.h>
+#include <vtkTrivialProducer.h>
+
+#include <vtkSMSourceProxy.h>
+
+#include "vtk_hdf5.h"
+
+#include <cassert>
+#include <string>
+#include <vector>
+
+#include <iostream>
+
+namespace tomviz {
+
+/**
+ * Write the data into the supplied group. This is really just mapping the C++
+ * vtkImageData types to the HDF5 API/types.
+ *
+ * dataTypeId refers to the type that will be stored in the HDF5 file
+ * memTypeId refers to the memory type that will be copied from vtkImageData
+ */
+template <typename T>
+bool writeVolume(T* buffer, hid_t groupId, const char* name, hid_t dataspaceId,
+                 hid_t dataTypeId, hid_t memTypeId)
+{
+  bool success = true;
+  hid_t dataId = H5Dcreate(groupId, name, dataTypeId, dataspaceId, H5P_DEFAULT,
+                           H5P_DEFAULT, H5P_DEFAULT);
+  if (dataId < 0) { // Failed to create object.
+    return false;
+  }
+  hid_t status =
+    H5Dwrite(dataId, memTypeId, H5S_ALL, H5S_ALL, H5P_DEFAULT, buffer);
+  if (status < 0) {
+    success = false;
+  }
+  status = H5Dclose(dataId);
+  if (status < 0) {
+    success = false;
+  }
+  return success;
+}
+
+class EmdFormat::Private
+{
+public:
+  Private() : fileId(H5I_INVALID_HID) {}
+  hid_t fileId;
+
+  hid_t createGroup(const std::string& group)
+  {
+    return H5Gcreate(fileId, group.c_str(), H5P_DEFAULT, H5P_DEFAULT,
+                     H5P_DEFAULT);
+  }
+
+  bool attribute(const std::string& group, const std::string& name, void* value,
+                 hid_t typeId)
+  {
+    if (H5Aexists_by_name(fileId, group.c_str(), name.c_str(), H5P_DEFAULT) <
+        0) {
+      // The specified attribute does not exist.
+      cout << group << name << " not found!" << endl;
```

Comment:
I may just remove these, or store a list of errors that we could put in the GUI. I don't have any Qt includes in the format, and was hoping to leave them out in case I wanted to move them into VTK, but a qDebug isn't the end of the world if it helps.